### ISSUE TITLE
Добавить обработчик onPageInit

### ISSUE THREAD

**Issue by KiberInfinity** [VkOpt/VkOpt#359]
При Ajax-навигации, у вк часто сначала меняется URL страницы, а только потом происходит инит основного модуля отвечающего за открытый раздел (страница группы, страница профиля и т.д). 
В результате при срабатывании onLocation в объекте "cur" не оказывается данных о названии текущего модуля и его владельца.
В качестве костыля сейчас используется повторный вызов onLocation:
https://github.com/VkOpt/VkOpt/commit/0dd03973f0d1b6353afd39379d60080a12416f68
https://github.com/VkOpt/VkOpt/commit/f9275109c6a738aa26c22a94e3985ce156c59103
что не совсем адекватно.

Предлагаю просмотреть модули вк с которыми аналогичная ситуация, и во все такие модули, в метод init, добавить инъекцию, которая будет триггерить новый обработчик onPageInit, в который добавлять уже тот код вкопта, который требует наличия данных о запущенном модуле, а повторный вызов onLocation убрать и перенести в новый обработчик код требующий этого костыля (пока что это ссылка на статистику группы и список вики-страниц)

**Comment by OwnageIsMagic** [comment]
`wontfix` - символично )

**Comment by KiberInfinity** [comment]
:D

**Comment by RblSb** [comment]
@KiberInfinity по поводу onLocation, можно в нем как-то отследить полную загрузку страницы событием, чтобы не юзать setTimeout?

**Comment by KiberInfinity** [comment]
Да вот как раз из-за подобной ситуации мы в этом разделе. Для некоторых разделов onLocation достаточно, а для некоторых полной можно считать загрузку тогда, когда "init" метод модуля вк для конкретного раздела был выполнен (Profile.init, Groups.init, Public.init и т.д), ибо у вк грузится каркас, потом грузится пачка айфреймов, и каждый айфрейм по завершению свой кусок передаёт в нужно место в каркасе страницы.
Вызов ".init" тоже происходит после одной из таких айфреймовых передач.

**Comment by RblSb** [comment]
@KiberInfinity а, тогда отлично. Значит onPageInit будет работать как onLocation+onInit, только когда страница полностью загружена?
А страница отображается только после добавления всех нужных модулей с афреймов, или они по отдельности могут появлятся? Если второе, то было бы неплохо еще иметь возможность проверять готовность нужного модуля страницы, что-то вроде события onModuleInit, которое отдает id аргументом каждого загруженного модуля.

**Comment by KiberInfinity** [comment]
Нет. onInit работает один раз после загрузки основной прототипной страницы (открыли вкладку, или нажали "обновить" вкладку). При аякс-переходах в пределах этой страницы он не вызывается повторно.

**Comment by KiberInfinity** [comment]
вот onPageInit и хотел заюзать для того события, когда готов модуль страницы. onInit и так выполняется, когда вкладка уже готова при первой загрузке страницы.

но чтоб не лепить костыли, возможно можно будет обойтись одной инъекцией:
```javascript
Inj.After('ajax._framenext',
          /ajax\.framedata\s*=\s*false;/, 
          '; setTimeout(vk_glue.page_ready_handler,2);'
         )
```

**Comment by RblSb** [comment]
@KiberInfinity я имел ввиду, что onPageInit будет работать как после открытия вкладки/обновления страницы, так и при аякс-переходах, или только во втором случае? Или в onInit сейчас можно и так без setTimeout работать, а такая проблема только при onLocation? Тогда окей.

**Comment by KiberInfinity** [comment]
проблема при onLocation. т.к на момент вызова onInit код инициализации модуля уже выполнен (инит модуля в DOM инлайн-скриптом прописывается при первой загрузке)

**Comment by RblSb** [comment]
@KiberInfinity извиняюсь, как-то только с третьего раза в ответе разобрался. onLocation будет теперь вызываться только один раз, сразу после смены адреса? Такое событие где-то может пригодиться?

**Comment by KiberInfinity** [comment]
@RblSb да по хорошему нужно чтоб onLocation срабатывал только один раз после смены адреса, когда модуль уже запущен. Сейчас смотрел можно ещё в `TopSearch.init` инъекцию добавить, т.к практически всегда вк его вызвает в том же скрипте, где происходит инит модуля-страницы.

Т.е предлагаю сейчас если после смены адреса страницы cur.module == undefined (да и cur.oid и cur.gid надо чекать), то поставить таймаут ожидания в пару секунд, который будет продлеваться при вызове `ajax.framegot` (т.е страница ещё грузится), если же был вызов `TopSearch.init`, то страница загрузилась и можно вызывать onLocation.
P.S. в админке группы, в некоторых разделах вообще отсутствует cur.module

**Comment by OwnageIsMagic** [comment]
```js
Profile.init.toString()
function(e){
"[inj_label]ProfileinitsetTimeoutvkgluenavhandler2";
return extend(...)  ; setTimeout(vk_glue.nav_handler,2);}
```
Таймаут всё-равно сработает только на следующем тике, так что можно заменить `Inj.End` на `Inj.Before`
